feat: control dev mode with env var

VishnuSanal opened the pull request:
**Description**

This PR fixes #62

sansyrox commented on robyn/cli.py:
```diff
@@ -104,6 +105,12 @@ def start_app_normally(config: Config):
+    load_vars(project_root=os.path.dirname(os.path.abspath(config.file_path)))
+
+    if not config.dev:
+        config.dev = bool(os.getenv("ROBYN_DEV_MODE", False))
```
could you add some documentation for this?

sansyrox commented on robyn/__init__.py:
```diff
@@ -48,7 +47,6 @@ def __init__(
         self.config = config
         self.dependencies = dependencies
 
-        load_vars(project_root=directory_path)
```
@VishnuSanal , this is needed here only. We would still want to allow the variables being set in `python app.py`

VishnuSanal replied:
@sansyrox `cli#run` gets called before `Robyn#init`. we cannot access the env vars from `cli#run` if the env vars are loaded later!

sansyrox replied:
@VishnuSanal , that is okay. But how would you load vars when using `python app.py` ?

sansyrox commented on docs_src/src/pages/documentation/api_reference/robyn_env.mdx:
```diff
@@ -16,6 +16,9 @@ Batman wanted to configure the server through an environment file. Changing code
     - Default: `False`
     - Example: `ROBYN_BROWSER_OPEN=True`
+ - `ROBYN_DEV_MODE`: Configures the dev mode
+    - Default: `False`
```
One follow up.

If we change it to `ROBYN_DEV_MODE=True` during execution, does it change to dev mode?

sansyrox commented on robyn/argument_parser.py:
```diff
@@ -23,7 +23,7 @@ def __init__(self) -> None:
             dest="dev",
             action="store_true",
-            default=False,
+            default=None,
```
Let's document this decision?

VishnuSanal replied:
@sansyrox It will be defaulted to false if an environment variable is not found. (since the fallback value to `os.getenv` on `cli#run` is false) Should we still document it?

sansyrox replied:
@VishnuSanal , I want you to document the decision that the env variable will have a priority over the cli flag. 

I believe we should prioritise cli flag over the env variable.

sansyrox commented:
Also, could you add a way to control the dev mode in `app.start(debug=True)`

sansyrox commented:
> the decision to turn on/off the dev mode would have been made before we reach robyn#start

And why is that? I thought you were implementing `python app.py --dev`

VishnuSanal commented:
> > the decision to turn on/off the dev mode would have been made before we reach `robyn#start`
> 
> And why is that?

what I meant with "the decision to turn on/off the dev mode would have been made before we reach `robyn#start`" is that that the control flow reaches `robyn#start` from `cli#run => cli#start_dev_server` or `cli#run => cli#start_app_normally`.

> I thought you were implementing `python app.py --dev` And even if it is `robyn app.py` Why is it being invoked before start?

I didn't get you? yeah, I am implementing the toggle for dev mode from env vars; but, in both cases, wouldn't the decision to switch to whether to start app normally or to start the dev server would have been made before the init?

sansyrox commented on robyn/__init__.py:
```diff
@@ -59,6 +61,10 @@ def __init__(
         if self.config.dev:
             exit("Dev mode is not supported in the python wrapper. Please use the CLI. e.g. python3 -m robyn app.py --dev ")
 
+        if not dev_mode_env_var and os.getenv("ROBYN_DEV_MODE", False) == "True":
```
Could you explain this line?

VishnuSanal replied:
here, I am checking whether the app is launched as a module or CLI. if it is CLI, the env var would be already set (before we call `load_vars` form `robyn#init` (hence, I am storing the value of `dev_mode_env_var` before calling `load_vars`.

sansyrox commented on robyn/cli.py:
```diff
@@ -104,6 +105,15 @@ def start_app_normally(config: Config):
 def run():
     config = Config()
+
+    if not config.file_path:
+        config.file_path = f"{os.getcwd()}/{__name__}"
```
is it needed?

VishnuSanal replied:
yes, otherwise the file path will be `None`, as we discussed in the previous meeting.

sansyrox commented on robyn/__init__.py:
```diff
@@ -48,6 +48,8 @@ def __init__(
         self.config = config
         self.dependencies = dependencies
 
+        dev_mode_env_var = os.getenv("ROBYN_DEV_MODE", False) == "True"
```
`__file__, file_object, __name__`

Try using these variables if possible

VishnuSanal replied:
```
print(__file__)
print(__name__)
print(file_object)
```

```
(.venv) RobynApp $ python3 app.py
.venv/lib/python3.12/site-packages/robyn/__init__.py
robyn
app.py

(.venv) RobynApp $ robyn app.py
.venv/lib/python3.12/site-packages/robyn/__init__.py
robyn
app.py

(.venv) RobynApp $ robyn -m src
.venv/lib/python3.12/site-packages/robyn/__init__.py
robyn
src/__main__.py

(.venv) RobynApp $ python3 -m src
.venv/lib/python3.12/site-packages/robyn/__init__.py
robyn
src/__main__.py

```

nops, no help :(

sansyrox replied:
@VishnuSanal , I have made an update to solve it.

sansyrox commented:
@VishnuSanal , updated the PR. Do have a look.

sansyrox commented:
@VishnuSanal , could you do a self review/review on my changes? And then I can do a final review 😄

VishnuSanal reviewed (commented):
it doesn't work as expected in my testing.

```
(.venv) RobynApp $ python3 app.py
INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
INFO:robyn.logger:Added route HttpMethod.GET /
INFO:robyn.logger:Robyn version: 0.56.1
INFO:robyn.logger:Starting server at http://0.0.0.0:8080
INFO:actix_server.builder:starting 1 workers
INFO:actix_server.server:Actix runtime found; starting in Actix runtime
^CINFO:actix_server.server:SIGINT received; starting forced shutdown
INFO:actix_server.worker:shutting down idle worker
INFO:actix_server.accept:accept thread stopped
Aborted (core dumped)

(.venv) RobynApp $ robyn app.py
Starting dev server...
Reloading /home/vishnu/Robyn/RobynApp/app.py...
INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
INFO:robyn.logger:Added route HttpMethod.GET /
INFO:robyn.logger:Robyn version: 0.56.1
INFO:robyn.logger:Starting server at http://0.0.0.0:8080
INFO:actix_server.builder:starting 1 workers
INFO:actix_server.server:Actix runtime found; starting in Actix runtime
^CINFO:actix_server.server:SIGINT received; starting forced shutdown
INFO:actix_server.accept:accept thread stopped
INFO:actix_server.worker:shutting down idle worker

(.venv) RobynApp $ cat robyn.env
ROBYN_DEV_MODE=True
```

sansyrox commented:
What is unexpected here @VishnuSanal ?

VishnuSanal commented:
> What is unexpected here @VishnuSanal ?

I have the environemnt variable set to true, but no warning is shown when running `python3 app.py` (CMIIW).

sansyrox commented:
@VishnuSanal , is it actually running in dev mode?

VishnuSanal commented:
> @VishnuSanal , is it actually running in dev mode?

it wasn't actually running in dev mode + the warning wasn't being shown, even after setting the env var. Please refer to the logs attached.

> ```
> (.venv) RobynApp $ python3 app.py
> INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
> INFO:robyn.logger:Added route HttpMethod.GET /
> INFO:robyn.logger:Robyn version: 0.56.1
> INFO:robyn.logger:Starting server at http://0.0.0.0:8080
> INFO:actix_server.builder:starting 1 workers
> INFO:actix_server.server:Actix runtime found; starting in Actix runtime
> ^CINFO:actix_server.server:SIGINT received; starting forced shutdown
> INFO:actix_server.worker:shutting down idle worker
> INFO:actix_server.accept:accept thread stopped
> Aborted (core dumped)
> 
> (.venv) RobynApp $ robyn app.py
> Starting dev server...
> Reloading /home/vishnu/Robyn/RobynApp/app.py...
> INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
> INFO:robyn.logger:Added route HttpMethod.GET /
> INFO:robyn.logger:Robyn version: 0.56.1
> INFO:robyn.logger:Starting server at http://0.0.0.0:8080
> INFO:actix_server.builder:starting 1 workers
> INFO:actix_server.server:Actix runtime found; starting in Actix runtime
> ^CINFO:actix_server.server:SIGINT received; starting forced shutdown
> INFO:actix_server.accept:accept thread stopped
> INFO:actix_server.worker:shutting down idle worker
> 
> (.venv) RobynApp $ cat robyn.env
> ROBYN_DEV_MODE=True
> ```

sansyrox commented:
@VishnuSanal , have a look once.

VishnuSanal commented:
yeps, works as expected! attaching the logs for reference.

```
(.venv) testing_application $ cat robyn.env
ROBYN_DEV_MODE=True

(.venv) testing_application $ python3 app.py
Ignoring ROBYN_DEV_MODE environment variable. Dev mode is not supported in the python wrapper.
Dev mode is not supported in the python wrapper. Please use the Robyn CLI. e.g. python3 -m robyn app.py

(.venv) testing_application $ robyn app.py
Starting dev server...
Reloading the server
argumens
INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
INFO:robyn.logger:Added route HttpMethod.GET /
INFO:robyn.logger:Robyn version: 0.56.1
INFO:robyn.logger:Starting server at http://0.0.0.0:8080
INFO:actix_server.builder:starting 1 workers
INFO:actix_server.server:Actix runtime found; starting in Actix runtime
^CINFO:actix_server.server:SIGINT received; starting forced shutdown
INFO:actix_server.worker:shutting down idle worker
INFO:actix_server.accept:accept thread stopped
```

```
(.venv) testing_application $ cat robyn.env
ROBYN_DEV_MODE=False

(.venv) testing_application $ python3 app.py
INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
INFO:robyn.logger:Added route HttpMethod.GET /
INFO:robyn.logger:Robyn version: 0.56.1
INFO:robyn.logger:Starting server at http://0.0.0.0:8080
INFO:actix_server.builder:starting 1 workers
INFO:actix_server.server:Actix runtime found; starting in Actix runtime
^CINFO:actix_server.server:SIGINT received; starting forced shutdown
INFO:actix_server.worker:shutting down idle worker
INFO:actix_server.accept:accept thread stopped
Aborted (core dumped)

(.venv) testing_application $ robyn app.py
INFO:robyn.logger:SERVER IS RUNNING IN VERBOSE/DEBUG MODE. Set --log-level to WARN to run in production mode.
INFO:robyn.logger:Added route HttpMethod.GET /
INFO:robyn.logger:Robyn version: 0.56.1
INFO:robyn.logger:Starting server at http://0.0.0.0:8080
INFO:actix_server.builder:starting 1 workers
INFO:actix_server.server:Actix runtime found; starting in Actix runtime
^CINFO:actix_server.server:SIGINT received; starting forced shutdown
INFO:actix_server.worker:shutting down idle worker
INFO:actix_server.accept:accept thread stopped
```

VishnuSanal commented on robyn/__init__.py:
```diff
@@ -69,6 +73,18 @@ def __init__(
+
+        if env_dev_mode and not is_robyn:
+            logger.error("Ignoring ROBYN_DEV_MODE environment variable. Dev mode is not supported in the python wrapper.")
+            raise SystemExit("Dev mode is not supported in the python wrapper. Please use the Robyn CLI. e.g. python3 -m robyn app.py")
```
@sansyrox we probably shouldn't raise an exit here; this would require the user to go to the env file & update it, just to make the app run. how about warning them & proceeding?

sansyrox replied:
@VishnuSanal , the file should be used locally. The flags should be toggled through a menu. 

I'd want to keep it as is for now. Should there be some concerns, we can allow it to run 😄

sansyrox reviewed (approved):
LGTM 😄
